### PR TITLE
[PLATFORM-367]: [Bridge_Ex] Exponential retry schedule

### PR DESCRIPTION
**Pull request by cpiemontese**
https://prima-assicurazioni-spa.myjetbrains.com/youtrack/issue/PLATFORM-367

This PR implements exponential backoff retries. Moreover, we refactored some of the code & updated the dependencies

### CODE REVIEW

**Review by matteosister — DISMISSED** [review]

**Review by cottinisimone — APPROVED** [review]

**Review by mbusi — COMMENTED** [review]
A few questions:
- why the random component in the delay calculation?
- why is the starting delay not configurable?
- why is the retry strategy (fixed/exponential) not configurable?

**Comment by cpiemontese** [comment]
> * why the random component in the delay calculation?

AFAIK it's because if we implemented a deterministic exponential backoff we could incur in cases of collisions like it's described [here](https://en.wikipedia.org/wiki/Exponential_backoff#Collision_avoidance). I can't say for sure if it would actually be a problem in our case, but we still implemented it like this since it seems to be common practice

> * why is the starting delay not configurable?

I don't know, I think it never was and we just kept it that way. We could easily expose an option to the client

> * why is the retry strategy (fixed/exponential) not configurable?

See [here](https://github.com/primait/bridge_ex/pull/30#pullrequestreview-889418348).
[TLDR; we went the "try to set a standard" route instead of a "here's a super configurable library" route]

**Comment by mbusi** [comment]
> > * why the random component in the delay calculation?
> 
> AFAIK it's because if we implemented a deterministic exponential backoff we could incur in cases of collisions like it's described [here](https://en.wikipedia.org/wiki/Exponential_backoff#Collision_avoidance). I can't say for sure if it would actually be a problem in our case, but we still implemented it like this since it seems to be common practice

> 
> > * why is the starting delay not configurable?
> 
> I don't know, I think it never was and we just kept it that way. We could easily expose an option to the client
> 
> > * why is the retry strategy (fixed/exponential) not configurable?
> 
> See [here](https://github.com/primait/bridge_ex/pull/30#pullrequestreview-889418348). TLDR; we went the "try to set a standard" route instead of a "here's a super configurable library" route

The way I see it is we should set sensible defaults so the user doesn't need to pass/config anything to make it work 'properly' but she's still free to override the settings and tweek the behaviour to her needs

File: lib/graphql/retry.ex [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  Utils for handling retrying of functions.
+  """
+
+  @spec retry(
+          String.t(),
+          (any() -> {:error, String.t()} | {:ok, any()}),
+          Keyword.t()
+        ) :: {:ok, any()} | {:error, any()}
+  def retry(arg, fun, retry_options) do
+    delay = Keyword.fetch!(retry_options, :delay)
```
Comment (cottinisimone) [inline lib/graphql/retry.ex:12]: Siamo sicuri di voler panicare se non ci sono ste option?

File: lib/graphql.ex [hunk 2]
```diff
@@ -93,7 +89,19 @@ defmodule BridgeEx.Graphql do
         http_options = Keyword.merge(@http_options, Keyword.get(options, :options, []))
         http_headers = Map.merge(@http_headers, Keyword.get(options, :headers, %{}))
         max_attempts = Keyword.get(options, :max_attempts, @max_attempts)
-        retry_policy = Keyword.get(options, :retry_policy, fn _ -> true end)
+
+        user_retry_options = Keyword.get(options, :retry_options, [])
+
+        retry_options =
+          Keyword.merge(
+            [
```
Comment (cottinisimone) [inline lib/graphql.ex:97]: Questa parte qui non è un module attribute per la funzione che non è serializzabile (o roba del genere, non ricordo bene)?

File: CHANGELOG.md [hunk 3]
```diff
@@ -9,11 +9,14 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 
-- New `retry_policy` option to `call`: clients can customize which errors are retried by providing a function
+- New `retry_options` to `call`: clients can customize how handle a call retry (more details [here](./README.md#customizing-the-retry-options)))
```
Comment (cottinisimone) [inline CHANGELOG.md:12]:
```suggestion
- New `retry_options` to `call`: clients can customize how handle a call retry (more details [here](./README.md#customizing-the-retry-options))
```

File: lib/graphql/retry.ex (hunk 1 above)
Comment (cpiemontese) [reply lib/graphql/retry.ex]: Ok uso `get` che almeno non panica, però ho aggiunto un commentino chiarificatore

**Review by andreapavoni — DISMISSED** [review]

File: lib/graphql.ex [hunk 4]
```diff
@@ -58,6 +58,13 @@ defmodule BridgeEx.Graphql do
       @max_attempts Keyword.get(unquote(opts), :max_attempts, 1)
       @log_options Keyword.get(unquote(opts), :log_options)
 
+      if Keyword.has_key?(unquote(opts), :max_attempts) do
```
Comment (andreapavoni) [inline lib/graphql.ex:61]: I was going to suggest this in the previous review, but I wasn't thinking about a way to implement it. Nice!

**Review by andreapavoni — APPROVED** [review]

**Review by matteosister — APPROVED** [review]

**Pull request merged**